src/ast/nodes/Identifier.ts

```diff
@@ -115,6 +130,9 @@ export default class Identifier extends NodeBase implements PatternNode {
 
 	hasEffects(): boolean {
 		if (!this.deoptimized) this.applyDeoptimizations();
+		if (this.isPossibleTDZ()) {
+			return true;
```

Comment:
I have to disagree with you on this point. This PR demonstrates that TDZ and var-use-before-decl are the same problem - they just result in different sorts of errors. 

I believe that case is fixable with some work. It is notable that it can be resolved with another pass:
```
$ echo 'var u = x; var x = 1;' | rollup --silent
x; var x = 1;
$ echo 'var u = x; var x = 1;' | rollup --silent | rollup --silent

$ 
```
In an earlier unrelated issue I examined how rollup might accomplish this without a reparse but got hung up on unexpected mutable behavior as result of repeated calls involving HasEffectsContext.

Reply:
Ah, I see. First I though that the deoptimization via `getLiteralValueAtPath` is enough, but indeed it extends to all other functions as well. Basically what should happen is that before declaration, any identifier should behave like "UNKNOWN_EXPRESSION". Maybe the simple fix would be to replace `this.variable` with some `this.getValue` that returns either `this.variable` or `UNKNOWN_EXPRESSION`, or maybe entirely replace `this.variable` with `UNKNOWN_EXPRESSION` for TDZ values. Will need to think about this.
In any case, this would again make the second pass unnecessary as `UNKNOWN_EXPRESSION` does not have a side effect when accessed.
Also I think the access side effect for TDZ variables should be moved from `hasEffects` to `hasEffectsWhenAccessedAtPath`.

Reply:
This is fixed now